docs/proposals/runtime-client-server.md

```diff
@@ -0,0 +1,202 @@
+   +                              +                               +
+   |                              |                               |
+   +---------SyncPod------------->+                               |
+   |                              |                               |
+   |                              +-----------RunPod------------->+
+   |                              +<------------------------------+
+   |                              |                               |
+   |                              XXXXXXXXXXXX                    |
+   |                              |          X                    |
+   |                              |    NetworkPlugin.             |
+   |                              |       SetupPod                |
+   |                              |          X                    |
+   |                              XXXXXXXXXXXX                    |
+   |                              |                               |
+   | <-------Success--------------+                               |
+   |                              |                               |
+   +                              +                               +
+```
+
+Furthermore, Kubelet network plugin has no sense of dedicated container runtime currently, but new container runtime may not work with existing docker or rkt plugin. So the type of container runtime and the ID of pod will be added to kubelet network plugin interface.
+
+For monitoring, cAdvisor is default monitoring solution in kubelet. But cAdvisor only supports docker and it is not working with other container runtimes. So the type of container runtime will be added to cAdvisor Interface. Each new container runtime should add their own runtime plugins to cAdvisor. This could be implemented in a few steps:
+
+* Making cAdvisor itself support container runtime plugins
+* Develop new container runtime plugins for cAdvisor
+* Kubelet calls cAdvisor APIs with configured plugins
+
+**`RemoteRuntime` Interface**
+
+```go
+type RemoteRuntime interface {
+  Type() VersionType
+
+  // GetPods returns a list containers group by pods. The boolean parameter
+  // specifies whether the runtime returns all containers including those already
+  // exited and dead containers (used for garbage collection).
+  GetPods(all bool) ([]*Pod, error)
+  // KillPod kills all the containers of a pod
+  KillPod(spec *api.PodSpec, runningPod Pod) error
+  // Runpod starts a pod with all containers and volumes
+  RunPod(pod *PodMeta, spec *api.PodSpec, volumes kubecontainer.VolumeMap, options []kubecontainer.RunContainerOptions) error
+  // GetPodStatus retrieves the status of the pod, including the information of
+  // all containers in the pod. Clients of this interface assume the containers
+  // statuses in a pod always have a deterministic ordering (eg: sorted by name).
+  GetPodStatus(pod *PodMeta) (*api.PodStatus, error)
+
+  // PullImage pulls an image from the network to local storage using the supplied
```

Comment:
Looks reasonable if we move image gc to runtime itself. @dchen1107 what do think about this?